### DISCUSSION TITLE
MSE playback support

### DISCUSSION THREAD

**Discussion by ozdonner1** [bluenviron/mediamtx#4049] [Questions]
### Question

Can you add support for MSE playback? It allows to play with very low latency and even go2rtc already support it.

**Comment by aler9** [comment]
This has been already answered in https://github.com/bluenviron/mediamtx/discussions/2589#discussioncomment-10562465
MSE has higher latency than LL-HLS with preload hints, which we already support, and has limited browser support.

**Reply by ozdonner1** [reply]
Thanks for the quick reply @aler9 .
However, LL-HLS has an issue - it needs at least one full segment in the playlist - as discussed in HLS.JS community: https://github.com/video-dev/hls.js/issues/6840
This impact on the start latency as you have to push and download the full segment, vs MSE where you don't have this limitation.

**Reply by aler9** [reply]
@ozdonner1 as written in the discussion, there is no such requirement, since the segment can be replaced with a gap. The problem is another one, and is that at the moment MediaMTX / gohlslib requires at least one full segment before providing the playlist, and that's because it needs the segment in order to estimate the mean segment duration and fill EXT-X-TARGETDURATION. Related discussion and patch are here: https://github.com/bluenviron/gohlslib/pull/92

**Reply by ozdonner1** [reply]
@aler9 the gap workaround is not working well on iOS devices, so it is not a full solution. The other problem you raised is because you need to estimate the mean segment duration and fill EXT-X-TARGETDURATION, however this is not required for MSE and it can give much lower latency for starting a playback. Supporting this will give lower latency.

**Reply by aler9** [reply]
MSE does not even exist on iPhone iOS. In MediaMTX we support a limited set of protocols that are standardized and guarantee a wide compatibility.
MSE does not fit into this definition.

**Reply by ozdonner1** [reply]
@aler9 it exists on Safari. It is also possible to get the media over websocket and extend the AVPlayer to play the video. MSE gives a lot of flexibility to play video. For example, using webcodecs https://developer.chrome.com/docs/web-platform/best-practices/webcodec

The Media Source Extension (MSE) was published by W3C (https://www.w3.org/TR/2022/WD-media-source-2-20220921/) body, providing a low-level toolkit that gave more control for managing buffering and resolution for adaptive streaming. MSE is supported by all browsers and it is used widely over the web. 

Without supporting MSE, to get low latency playback you need to connect MediaMTX to a tool like go2rtc to get MSE which is much more complex setup for such a feature. LL-HLS is not solving the start latency issue and latency has a big user experience impact. 

We think really help to many users and it will be great for MediaMTX to support it. We are happy to help.
